File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
public int maxProduct(int[] nums) {
    if (nums == null || nums.length == 0) {
        return 0;
    }

    int maxProduct = nums[0];
    int maxEndingHere = nums[0];
    int minEndingHere = nums[0];

    for (int i = 1; i < nums.length; i++) {
        if (nums[i] < 0) {
            // Swap maxEndingHere and minEndingHere when encountering a negative number
            int temp = maxEndingHere;
            maxEndingHere = minEndingHere;
            minEndingHere = temp;
        }

        // Update the maximum and minimum products ending at the current position
        maxEndingHere = Math.max(nums[i], maxEndingHere * nums[i]);
        minEndingHere = Math.min(nums[i], minEndingHere * nums[i]);

        // Update the overall maximum product
        maxProduct = Math.max(maxProduct, maxEndingHere);
    }

    return maxProduct;
}

}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution(object):
    def maxProduct(self,nums):
        prod1 = nums[0]
        prod2 = nums[0]
        result = nums[0]

        for i in range(1, len(nums)):
            temp = max(nums[i], prod1 * nums[i], prod2 * nums[i])
            prod2 = min(nums[i], prod1 * nums[i], prod2 * nums[i])
            prod1 = temp

            result = max(result, prod1)

        return result
        